File: graph/number_of_islands.py
from typing import List
from collections import deque

class UnionFind:
    def __init__(self, total_elements):
        self.parent = list(range(total_elements))
        self.rank = [1] * total_elements
        self.count = 0  # Tracks number of islands

    def find(self, x):
        while x != self.parent[x]:  # Iterative path compression
            self.parent[x] = self.parent[self.parent[x]]  # Path compression
            x = self.parent[x]
        return x

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.parent[rootY] = rootX
            elif self.rank[rootX] < self.rank[rootY]:
                self.parent[rootX] = rootY
            else:
                self.parent[rootY] = rootX
                self.rank[rootX] += 1
            self.count -= 1  # Merge islands, reduce count

class Solution:
    def numIslands1(self, grid: List[List[str]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        visited = set()
        islands = 0
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        def dfs(r, c):
            if(r not in range(ROWS) or c not in range(COLS) or
            grid[r][c] != '1' or (r, c) in visited):
                return
            visited.add((r, c))
            for d in directions:
                dr, dc = r + d[0], c + d[1]
                dfs(dr, dc)


        for r in range(ROWS):
            for c in range(COLS):
                if((r, c) not in visited and grid[r][c] == '1'):
                    islands += 1
                    dfs(r, c)

        return islands

    def numIslands2(self, grid: List[List[str]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        visited = set()
        islands = 0
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        def bfs(row, col):
            q = deque()
            q.append([row, col])
            while q:
                r, c = q.popleft()
                if(r in range(ROWS) and c in range(COLS) and
            grid[r][c] == '1' and (r, c) not in visited):
                    visited.add((r, c))
                    for d in directions:
                        q.append([r + d[0], c + d[1]])

        for r in range(ROWS):
            for c in range(COLS):
                if((r, c) not in visited and grid[r][c] == '1'):
                    islands += 1
                    bfs(r, c)

        return islands

    def numIslands(grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0

        ROWS, COLS = len(grid), len(grid[0])
        uf = UnionFind(ROWS * COLS)

        # Count initial number of islands
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == '1':
                    uf.count += 1  # Every '1' is initially an island

        # Union adjacent land cells
        directions = [(0, 1), (1, 0), (-1, 0), (0, -1)]
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == '1':
                    index = r * COLS + c
                    for dr, dc in directions:
                        nr, nc = r + dr, c + dc
                        if 0 <= nr < ROWS and 0 <= nc < COLS and grid[nr][nc] == '1':
                            uf.union(index, nr * COLS + nc)

        return uf.count  # Number of connected components

File: graph/number_of_islands.java
class Solution {
    private int ROWS, COLS;
    private final int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    private Set<String> visited = new HashSet<>();

    public int numIslands1(char[][] grid) {
        if(grid.length == 0 || grid[0].length == 0){
            return 0;
        }
        ROWS = grid.length;
        COLS = grid[0].length;
        int islands = 0;
        
        for(int r = 0; r < ROWS; r++){
            for(int c = 0; c < COLS; c++){
                if(grid[r][c] == '1' && !visited.contains(r + "-" + c)){
                    islands++;
                    dfs(grid, r, c);
                }
            }
        }

        return islands;

    }

    public void dfs(char[][] grid, int r, int c){
        if(r < 0 || r == ROWS || c < 0 || c == COLS || grid[r][c] != '1' || visited.contains(r + "-" + c)){
            return;
        }

        visited.add(r + "-" + c);
        for(int[] dir : directions){
            dfs(grid, r + dir[0], c + dir[1]);
        }

    }

    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0 || grid[0].length == 0) {
            return 0;
        }

        int rows = grid.length;
        int cols = grid[0].length;
        int islands = 0;
        Set<String> visited = new HashSet<>();

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == '1' && !visited.contains(r + "-" + c)) {
                    islands++;
                    bfs(grid, r, c, visited);
                }
            }
        }

        return islands;
    }

    private void bfs(char[][] grid, int r, int c, Set<String> visited) {
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{r, c});
        visited.add(r + "-" + c);

        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            for (int[] dir : directions) {
                int newRow = curr[0] + dir[0];
                int newCol = curr[1] + dir[1];
                if (newRow >= 0 && newRow < grid.length && newCol >= 0 && newCol < grid[0].length && grid[newRow][newCol] == '1' && !visited.contains(newRow + "-" + newCol)) {
                    queue.offer(new int[]{newRow, newCol});
                    visited.add(newRow + "-" + newCol);
                }
            }
        }
    }
}